src/main/java/fr/sncf/osrd/speedcontroller/generators/MarecoAllowanceGenerator.java

```diff
@@ -106,29 +131,95 @@ protected double getFirstHighEstimate() {
 
     @Override
     protected double getFirstGuess() {
-        return this.getFirstHighEstimate() / (1 + value / 100);
+        return (this.getFirstHighEstimate() + this.getFirstLowEstimate()) / 2;
     }
 
-    private List<Double> findPositionSameSpeedAsVF(SortedDoubleMap speeds, double vf) {
-        // TODO check only in deceleration intervals
-        boolean isLastSpeedBelowVF = true;
+    /** Finds the end position of the coasting phases, that will then be generated backwards
+     * with generateCoastingSpeedControllerAtPosition()
+     * Coasting phases can be generated for two different reasons
+     * - a deceleration phase where braking will be replaced by coasting
+     * - before an accelerating slope */
+    private List<Double> findEndOfCoastingPositions(SortedDoubleMap speeds, double v1) {
         var res = new ArrayList<Double>();
-        for (var position : speeds.navigableKeySet()) {
-            var speed = speeds.get(position);
-            boolean isCurrentSpeedBelowVF = speed < vf;
-            if (isCurrentSpeedBelowVF && !isLastSpeedBelowVF && isDecelerating(position)) {
-                res.add(position);
+        var rollingStock = schedule.rollingStock;
+        double vf = vf(v1);
+        var limitAnnounceSpeedControllers = findLimitSpeedAnnouncers(maxSpeedControllers);
+
+        // coasting before deceleration phases
+        for (var announcer : limitAnnounceSpeedControllers) {
+            // if that LimitAnnounceSpeedController is above v1 that means it will not have an impact here
+            if (v1 >= announcer.targetSpeedLimit) {
```

Comment:
You can reverse the condition and continue. This will remove a level of indentation.

Reply:
Good idea, I'll do it.